nav2_util/include/nav2_util/twist_publisher.hpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (C) 2023 Ryan Friedman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef NAV2_UTIL__TWIST_PUBLISHER_HPP_
+#define NAV2_UTIL__TWIST_PUBLISHER_HPP_
+
+
+#include <memory>
+#include <string>
+
+#include "geometry_msgs/msg/twist.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_lifecycle/lifecycle_publisher.hpp"
+#include "rclcpp/qos.hpp"
+
+namespace nav2_util
+{
+
+/**
+ * @class nav2_util::TwistPublisher
+ * @brief A simple wrapper on a Twist publisher that provides either Twist or TwistStamped
+ *
+ * The default is to publish Twist to preserve backwards compatibility, but it can be overridden
+ * using the "enable_stamped_cmd_vel" parameter to publish TwistStamped.
+ *
+ */
+
+class TwistPublisher
+{
+public:
+  /**
+  * @brief A constructor
+  * @param nh The node
+  * @param topic publisher topic name
+  * @param qos publisher quality of service
+  */
+  explicit TwistPublisher(
+    nav2_util::LifecycleNode::SharedPtr nh,
+    const std::string & topic,
+    const rclcpp::QoS & qos
+  )
+  : topic_(topic), node_(nh)
+  {
+    assert(nh != nullptr);
+    node_->declare_parameter("enable_stamped_cmd_vel", false);
+    node_->get_parameter("enable_stamped_cmd_vel", is_stamped_);
+    if (is_stamped_) {
+      twist_stamped_pub_ = node_->create_publisher<geometry_msgs::msg::TwistStamped>(
+        topic_,
+        qos);
+    } else {
+      twist_pub_ = node_->create_publisher<geometry_msgs::msg::Twist>(
+        topic_,
+        qos);
+    }
+  }
+
+
+  void on_activate()
+  {
+    if (is_stamped_) {
+      twist_stamped_pub_->on_activate();
+    } else {
+      twist_pub_->on_activate();
+    }
+  }
+
+  void on_deactivate()
+  {
+    if (is_stamped_) {
+      twist_stamped_pub_->on_deactivate();
+    } else {
+      twist_pub_->on_deactivate();
+    }
+  }
+
+  [[nodiscard]] bool is_activated() const
+  {
+    if (is_stamped_) {
+      return twist_stamped_pub_->is_activated();
+    } else {
+      return twist_pub_->is_activated();
+    }
+  }
+
+  void publish(const geometry_msgs::msg::TwistStamped & velocity)
```

Comment:
A non-stamped version that adds a stamp if `is_stamped_` is true? Basically the counter-implementation of this one

Reply:
I was thinking about this, and was planning to add it when it's needed, however it seems a lot of NAV2 internals are using the stamped version anyways.

Reply:
Oh ok, then I don't think its strictly necessary then.

The only thing I'd say in counter to that is that if others want to use this to stamp their twists (with a param/method arg for the base_frame) that could be a nice add on. But given your expertise / applications i think you're in the best position to tell me if that's even something desired. If not, feel free to ignore.